### PR TITLE
tests: subsys: shell: add fuzzer configuration for shell commands

### PR DESCRIPTION
**Pull request by nordic-piks**
Allow to run tests with libFuzzer.

### CODE REVIEW

**Comment by nordic-piks** [comment]
There are already some found issues, listing them here until shell developers do confirm valid test setup:
```
zephyr/subsys/shell/shell_history.c:158:16: runtime error: applying zero offset to null pointer
    #0 0x5a7705 in z_shell_history_put zephyr/subsys/shell/shell_history.c:158:16
    #1 0x587605 in history_put zephyr/subsys/shell/shell.c:189:2
    #2 0x587605 in execute zephyr/subsys/shell/shell.c:649:3
    #3 0x586d95 in shell_execute_cmd zephyr/subsys/shell/shell.c:1655:12
    #4 0x54e94e in _posix_zephyr_main zephyr/tests/subsys/shell/shell_fuzzer/build/../src/main.c:64:9

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior zephyr/subsys/shell/shell_history.c:158:16 in
MS: 2 ChangeByte-Custom-; base unit: 324e10e8df6541c20978387bd7b2845ba5dbf81e
0xa,0xa,0x76,0x0,0x0,0x0,0xa,0xa,0x0,0x0,0x0,0x0,0x0,0xa,0xa,0x0,
\x0a\x0av\x00\x00\x00\x0a\x0a\x00\x00\x00\x00\x00\x0a\x0a\x00
artifact_prefix='./'; Test unit written to ./crash-049243cbb565af38ab0e2c4851606ad770078f43
Base64: Cgp2AAAACgoAAAAAAAoKAA==
```
(suppression: `pointer-overflow:z_shell_history_put`)

```
zephyr/subsys/shell/shell_history.c:82:11: runtime error: applying zero offset to null pointer
    #0 0x5a66e7 in z_shell_history_get zephyr/subsys/shell/shell_history.c:82:11
    #1 0x5a2d10 in cmd_history zephyr/subsys/shell/shell_cmds.c:365:3
    #2 0x589237 in exec_cmd zephyr/subsys/shell/shell.c:558:13
    #3 0x589237 in execute zephyr/subsys/shell/shell.c:800:9
    #4 0x586d95 in shell_execute_cmd zephyr/subsys/shell/shell.c:1655:12
    #5 0x54e94e in _posix_zephyr_main zephyr/tests/subsys/shell/shell_fuzzer/build/../src/main.c:64:9

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior zephyr/subsys/shell/shell_history.c:82:11 in
MS: 8 InsertRepeatedBytes-Custom-ChangeByte-Custom-CMP-Custom-CMP-Custom- DE: "Y\x00"-"history"-; base unit: 44d37eca4ef65d793f7bfe0778716537d73141cc
0x68,0x69,0x73,0x74,0x6f,0x72,0x79,0x0,0x0,0x0,0x0,0x0,0x19,0x20,0x0,
history\x00\x00\x00\x00\x00\x19 \x00
artifact_prefix='./'; Test unit written to ./crash-fa2d4be21c475dba9e19378072b281620da0038e
Base64: aGlzdG9yeQAAAAAAGSAA
```
(suppression: `pointer-overflow:z_shell_history_get`)


```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==3065==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x0000005758eb bp 0x7fcd96564890 sp 0x7fcd96564860 T3)
==3065==The signal is caused by a READ memory access.
==3065==Hint: address points to the zero page.
    #0 0x5758eb in sys_read32 zephyr/tests/subsys/shell/shell_fuzzer/build/../../../../../include/zephyr/arch/common/sys_io.h:45:9
    #1 0x5758eb in memory_read zephyr/subsys/shell/modules/devmem_service.c:271:11
    #2 0x5758eb in cmd_devmem zephyr/subsys/shell/modules/devmem_service.c:339:10
    #3 0x5892eb in exec_cmd zephyr/subsys/shell/shell.c:558:13
    #4 0x5892eb in execute zephyr/subsys/shell/shell.c:800:9
    #5 0x586e49 in shell_execute_cmd zephyr/subsys/shell/shell.c:1655:12
    #6 0x54e94e in _posix_zephyr_main zephyr/tests/subsys/shell/shell_fuzzer/build/../src/main.c:63:9


==3065==ABORTING
MS: 2 CopyPart-Custom-; base unit: c4a35c86f3438edb95dbef2471b3fb473f1099f4
0x64,0x65,0x76,0x6d,0x65,0x6d,0x9,0x6d,0x0,
devmem\x09m\x00
artifact_prefix='./'; Test unit written to ./crash-11b4f88718948f0807d717cf8b01b22c86d6438d
Base64: ZGV2bWVtCW0A
```

File: tests/subsys/shell/shell_fuzzer/src/main.c [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+size_t LLVMFuzzerCustomMutator(uint8_t *Data, size_t Size, size_t MaxSize, unsigned int Seed)
+{
+	Size = LLVMFuzzerMutate(Data, Size, MaxSize);
+	Data[Size - 1] = 0;
+	return Size;
+}
+
+int main(void)
+{
+	printk("Hello World! %s\n", CONFIG_BOARD);
```
Comment (nordic-krch) [inline tests/subsys/shell/shell_fuzzer/src/main.c:44]: Maybe "shell libfuzzer test" instead?

File: tests/subsys/shell/shell_fuzzer/src/main.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		if (posix_fuzz_sz == 0 || posix_fuzz_sz == 1) {
+			/* discard 0 size vectors as it is no a valid string
+			 * discard 1 size vectors as they do not have termination
+			 */
+			continue;
+		}
+		const struct shell *sh = shell_backend_dummy_get_ptr();
+
+		/* Wait for the initialization of the shell dummy backend. */
+		WAIT_FOR(shell_ready(sh), 20000, k_msleep(1));
```
Comment (nordic-krch) [inline tests/subsys/shell/shell_fuzzer/src/main.c:64]: can that be moved before the loop?

File: tests/subsys/shell/shell_fuzzer/src/main.c (hunk 1 above)
Comment (nordic-piks) [reply tests/subsys/shell/shell_fuzzer/src/main.c]: Done, thanks :)

**Comment by nordic-piks** [comment]
CC: @nashif There is interesting approach for fuzzing testing. Please have a look.
At future, if this topic raise more interest, twister could possibly handle fuzzing test execution - currently, there is instruction how to run it manually - README.rst.

File: tests/subsys/shell/shell_fuzzer/testcase.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+tests:
+  shell.fuzzer:
+    tags: shell fuzzer
```
Comment (hakehuang) [inline tests/subsys/shell/shell_fuzzer/testcase.yaml:3]: it depends on the CONFIG_ARCH_POSIX_FUZZ_IRQ, so need add 
filter:  CONFIG_ARCH_POSIX_FUZZ_IRQ

**Comment by MaureenHelm** [comment]
@jakub-uC can you take a look please?

**Comment by nordic-piks** [comment]
@jakub-uC Any comment on that PR? on those issues?

**Comment by nordic-piks** [comment]
Closing due to no interest in this topic (at least in the shell area).
If there is another area - more suitable for such testing, I will continue this work.

**Comment by jakub-uC** [comment]
@nordic-piks  sorry I've missed it

**Comment by nordic-piks** [comment]
@jakub-uC  You can still look on code and reported issues ;)

**Review by jakub-uC — APPROVED** [review]

**Pull request closed without merging**